### DISCUSSION TITLE
Custom Nordic nRF52840 Zigbee Device - Connecting to Zigbee2MQTT

### DISCUSSION THREAD

**Discussion by rechichidaniel** [Koenkk/zigbee2mqtt#16812] [General]
Has anyone been able to get the Nordic [Zigbee Light Bulb](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/zigbee/light_bulb/README.html) or [Zigbee Template](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/zigbee/template/README.html) example projects to connect to Zigbee2MQTT yet?

I realise I will need to write a custom converter to make this work, but I haven't made it that far yet because I'm still trying to get the `nRF52840 DK` to connect to the Z2M network.

According to [this Nordic forum post](https://devzone.nordicsemi.com/f/nordic-q-a/75186/join-non-nrf-network), it should be possible as long as the Z2M Zigbee stack follows the Zigbee spec, mainly as a secondary point to why the dev kit can't connect to ZHA's zigpy.

My device is trying to connect, but just keeps printing `I: Network steering was not successful (status: -1)`.

**Comment by rechichidaniel** [comment]
The problem was that the sample light bulb code was defaulting to channel 16, and not searching for the channel I was using.
My Nordic nRF52840DK is now connected to my Zigbee2MQTT network.
https://devzone.nordicsemi.com/f/nordic-q-a/97235/nrf52-sdk-2-2-0-change-default-zigbee-channel-and-join-existing-non-nrf-network

**Reply by rechichidaniel** [reply]
I'm sure most people wouldn't need this, but here is my converter.

Notes:
- `Power on Behavior` doesn't work, but that's because it's not built into the Nordic light bulb sample.
- The `Dimable_Light_v0.1`  zigbeeModel is straight from the SDK `main.c` sample code. It's easy to change.

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const utils = require('zigbee-herdsman-converters/lib/utils');
const globalStore = require('zigbee-herdsman-converters/lib/store');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['Dimable_Light_v0.1'],
    model: 'nRF52840DK',
    vendor: 'Nordic',
    description: 'Nordic nRF52840DK, SDK V2.2.0, Light Bulb Sample',
    extend: extend.light_onoff_brightness(),
};

module.exports = definition;
```

**Reply by junosuarez** [reply]
Thanks for sharing! I'm interested in getting started making my own DIY zigbee devices (whenever shipping from AliExpress gets here, lol) - so any examples are more than welcome.

**Reply by rechichidaniel** [reply]
Thanks Juno,

I'm new to this too. I've spent the last few months learning about clusters and commands, but now my challenge is trying to figure out what in the nRF SDK I need to edit in order to make the dev kit use an external LED driver instead of its onboard LEDs.

**Comment by jj-uk** [comment]
Have you made any progress with this?

I'd like to make a device that can be controlled from z2m, send and receive data etc. 

The sdk templates aren't that great.

**Reply by rechichidaniel** [reply]
The SDK samples are terrible, and the whole SDK is a mess.
I responded to your question but accidentally left a new comment instead of replying.
Sorry, you should be able to see it below.

**Reply by nobodyguy** [reply]
Minimal set of ZigBee clusters is included in the zboss lib shipped with nrf connect SDK. The missing pieces have to be implemented according to the ZCL spec.
I implemented CO2 concentration measurement cluster here https://github.com/nobodyguy/zigbee_air_quality_monitor_firmware

**Comment by rechichidaniel** [comment]
Thanks jj-uk. Yep, I've made some progress but not a lot. I can at least get the light bulb and template samples connected.

### nRF Connect SDK Changes

By default, the nRF SDK samples only search for ZigBee channel 16, presumably to speed up the scan time. you have to change that to whatever channel you're using or make them scan every channel.
This is where I posted to learn that: [nRF52 SDK 2.2.0 change default Zigbee channel and join existing non-nRF network](https://devzone.nordicsemi.com/f/nordic-q-a/97235/nrf52-sdk-2-2-0-change-default-zigbee-channel-and-join-existing-non-nrf-network)

To make them scan every channel, you can either edit prj.conf using the Kconfig GUI, or you can add this like to the file directly.
```
CONFIG_ZIGBEE_CHANNEL_SELECTION_MODE_MULTI=y
```
This is what my Kconfig GUI looks like after making that prj.conf change.
By default, `802.15.4 channel used by Zigbee <11-26>` is set to 16, and the `Mask of 802.15.4 channels used by Zigbee` is empty.
![image](https://github.com/Koenkk/zigbee2mqtt/assets/21980408/4dcf365c-c7df-4025-b993-6be97bc0ca15)

Pay close attention to what you set your model ID to. You'll need this in your Z2M converter. Note that `BULB_INIT_BASIC_MODEL_ID        ` can have a different name in each sample.
```
/* Model number assigned by manufacturer (32-bytes long string). */
#define BULB_INIT_BASIC_MODEL_ID        "Dimable_Light_v0.1"
```




### Zigbee2MQTT Changes

The converter I posted in my previous comment works.
https://github.com/Koenkk/zigbee2mqtt/discussions/16812#discussioncomment-5189239
Although, make sure you update `zigbeeModel`, `model`, `vendor` and `description` to whatever you need them to be.
**`zigbeeModel` is the important one and needs to match what you set in your nRF Connect code.**
The `extend` you choose depends on your device. However, using an `extend` doesn't give you a lot of freedom to add or remove features because it's predefined. You can create your own extend function within the converter, but I recommend using the `fromZigbee`, `toZigbee` and `exposes` options so that you can have more control in a simpler way.

e.g.
```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const utils = require('zigbee-herdsman-converters/lib/utils');
const globalStore = require('zigbee-herdsman-converters/lib/store');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['Dimable_Light_v0.1'],
    model: 'nRF52840DK',
    vendor: 'Nordic',
    description: 'Nordic nRF52840DK, SDK V2.2.0, Light Bulb Sample - exposes',
    
    fromZigbee: [fz.on_off, fz.brightness, fz.level_config, fz.ignore_basic_report, fz.temperature],
    toZigbee: [tz.light_onoff_brightness, tz.ignore_rate, tz.level_config],
    exposes: [e.light_brightness(), e.temperature()],
    // e.temperature().withEndpoint(12)
    
};

module.exports = definition;
```

**Comment by Hedda** [comment]
FYI, check out this new open-source firmware project by ffenix113 which aims to (eventually) provide similar functionality to [ESPHome](https://esphome.io/) but allow users to make their own Zigbee devices based on nRF52840 (Nordic Semi nRF series) but note that it is still in very early development:

https://github.com/ffenix113/zigbee_home

https://ffenix113.github.io/zigbee_home/

https://github.com/Koenkk/zigbee2mqtt/discussions/21029

**Reply by jj-uk** [reply]
This looks interesting!  I'll have to take a look. I just can't get my head around visual code or the nordic sdk so gave up a while ago and looked at the STM32WB55 devices which were 'very' easy to use. Creating a working project was fast, the IDE does most of the code generation for you, but it's just a shame that the ST modules are expensive and difficult to use (PCB layout requirements) when compared to the nRF-series modules.

Looking forward to the zigbee-home project.

**Reply by Hedda** [reply]
Suggest you post any questions, suggestions, or ideas, etc. there for feedback and input:

https://github.com/ffenix113/zigbee_home/issues

https://github.com/ffenix113/zigbee_home/discussions
